### DISCUSSION TITLE
Bake Constraints to Keymesh

### DISCUSSION THREAD

**Discussion by Dereuz24** [nickberckley/keymesh#39] [Feature Requests]
**Describe the feature you'd like to see**

Provide a setting allowing you to select between baking to keymesh the object animation (action, armature, shapekey) and/or the constraints on the object.

**Comparisons**

There are many cases in which an object is constrained to an armature or other objects with any number of transforms or relationship constraints for convenience. 

Instead of weighting an object to a bone and animating this bone, we simply use copy location/rotation, armature constraints, action constraints, pivots, shrinkwrap, track-to, etc directly on objects, later these properties are keyframed and animated in the timeline.

Examples: 

- A weapon changes parenting between characters.
- A vehicle is constrained to a curve to follow a path
- An armor on a character has an armature constraint on a bone of the character instead of using an armature modifier.

To obtain the final visual transforms of any arbitrary number of constraints  these must be applied in order like modifiers (from top to bottom) per selected object, per frame. This way you can obtain an object final visual transform/animation in the viewport by baking down these relationships.

The reason I say there should be a checkbox or a setting in the "bake to keymesh" button is because there's are situations I see these constraints being useful by providing a non destructive way to move the objects around, but it depends on the application and situation.

**Comment by nickberckley** [comment]
Interesting, thanks for giving examples. Constraints are something I've already thought about and are on the to-do in #27 but I didn't include them in the initial version because I don't have enough knowledge about them to know what the expected outcome is. The situation you point out is a good example, but I would still need help with some of the topics.

The biggest issue is that constraints aren't modifying the mesh, they're modifying the object. Keymesh doesn't really touch objects now, only object data. When constraints are applied they actually change object transformations, instead of modifying mesh (like in edit mode). Applying them frame-by-frame would mean animating object transforms. This has some implications:

1. If object transforms are already animated applying constraint will add to animated values and keyframe new sum values. This will drastically change how f-curves and dope sheet look, and might make the operator useless because user will not be able to animate object transforms anymore (except frame-by-frame).
2. Operator can potentially animate delta transforms instead, but is it desirable? What if delta transforms are animated too? Another issue as above.
3. How to get which transforms to keyframe? If the operator simply keyframes all three on each frame it will create a lot of unnecessary keyframes for unchanged values. Calculating which transform needs to be keyframed might not be easy, and require whole lot of "smart" code to figure out when what is keyframed.

Another big question is the order of baking (when everything is baked at once). I assume that both of them can affect each other and produce unexpected results:
- Are modifiers (and shape keys) applied first and constraints second?
- Are constraints applied before modifiers & shape keys?
- Should this be operator property?

If you can provide .blend files that have constraints like you described it would help the most for testing. But in general, this isn't something that can come quickly because it's not easy to implement and also I will not be spending as much time on add-on for a while.
